src/config/watcher.rs

```diff
@@ -19,11 +22,40 @@ const CONFIG_WATCH_DELAY: std::time::Duration = std::time::Duration::from_secs(1
 
 const RETRY_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
 
+pub enum WatcherConfig {
+    /// recommended watcher for os, usually inotify for linux based systems
+    RecommendedWatcher,
+    /// poll based watcher. for watching files from NFS.
+    PollWatcher(u64),
+}
+
+enum Watcher {
+    /// recommended watcher for os, usually inotify for linux based systems
+    RecommendedWatcher(RecommendedWatcher),
+    /// poll based watcher. for watching files from NFS.
+    PollWatcher(PollWatcher),
+}
+
+impl Watcher {
+    fn add_paths(&mut self, config_paths: &[PathBuf]) -> Result<(), Error> {
+        let watcher: &mut dyn NotifyWatcher = match self {
```

Comment:
you can get rid of the dynamic dispatch if you return the `Watcher` enum here.

You can implement the trait `From<RecommendedWatcher> for Watcher` and `From<PollingWatcher> Watcher` to simplify the creation of those here. You could then just call `.into`

Read this if you are not familiar with that https://doc.rust-lang.org/std/convert/trait.From.html

Reply:
You can then implement a `watch` method in `Watcher` that relies on the underlying `notify::Watcher` variants

Reply:
great i'll do this.

Reply:
I just added the watch method for `Watcher`. that eliminates the dynamic dispatch. 
can u check if this is okay